feat: change names of groups from Priority Field and change the sort of order groups same as task sorting order

ilandikov opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

# Description

- change the naming of groups for the Priority field
- change the sort of order of grouping for all fields to sort by task sorting order instead of by group name alphabetical order

## Motivation and Context

Fixes #1628 . Priority groups are now displayed without numbers: `High`, `Medium`, `None`, `Low`, but are sorted in the same order and not in alphabetical order (`High`, `Low`, `Medium`, `None`).

## How has this been tested?

<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, tests ran to see how -->
<!--- your change affects other areas of the code, etc. -->

## Screenshots (if appropriate)

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

Changes visible to users:

- [ ] **Bug fix** (prefix: `fix` - non-breaking change which fixes an issue)
- [x] **New feature** (prefix: `feat` - non-breaking change which adds functionality)
- [ ] **Breaking change** (prefix: `feat!!` or `fix!!` - fix or feature that would cause existing functionality to not work as expected)
- [ ] **Documentation** (prefix: `docs` - improvements to any documentation content **for users**)
- [ ] **Sample vault** (prefix: `vault` - improvements to the [Tasks-Demo sample vault](https://github.com/obsidian-tasks-group/obsidian-tasks/tree/main/resources/sample_vaults/Tasks-Demo))
- [ ] **Contributing Guidelines** (prefix: `contrib` - any improvements to documentation content **for contributors** - see [Contributing to Tasks](https://publish.obsidian.md/tasks-contributing/))

Internal changes:

- [ ] **Refactor** (prefix: `refactor` - non-breaking change which only improves the design or structure of existing code, and making no changes to its external behaviour)
- [ ] **Tests** (prefix: `test` - additions and improvements to unit tests and the smoke tests)
- [ ] **Infrastructure** (prefix: `chore` - examples include GitHub Actions, issue templates)

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] My code follows the code style of this project and passes `yarn run lint`.
- [ ] My change requires a change to the documentation.
- [ ] I have [updated the documentation](https://publish.obsidian.md/tasks-contributing/Documentation/About+Documentation) accordingly.
- [x] My change has adequate [Unit Test coverage](https://publish.obsidian.md/tasks-contributing/Testing/About+Testing).

## Terms

<!--
By submitting this pull request, you must agree to follow our
[contributing guide](https://publish.obsidian.md/tasks-contributing) and
[Code of Conduct](https://github.com/obsidian-tasks-group/obsidian-tasks/blob/main/CODE_OF_CONDUCT.md).
Put an x in the boxes to confirm you agree.
-->

- [x] My contribution follow this project's [contributing guide](https://publish.obsidian.md/tasks-contributing)
- [x] I agree to follow this project's [Code of Conduct](https://github.com/obsidian-tasks-group/obsidian-tasks/blob/main/CODE_OF_CONDUCT.md)

claremacrae commented:
How does this relate to PR #1943?
I mean, I think there will be conflicts whichever order this one and that one are merged.

claremacrae commented:
From the PR description it's not obvious if this is talking about renaming the priorities themselves. Something more specific would be really useful.

claremacrae commented:
> How does this relate to PR #1943?
> I mean, I think there will be conflicts whichever order this one and that one are merged.

If this one is going to go ahead, please do update the user docs, so that a complete PR can be reviewed. Many thanks.

ilandikov commented:
> How does this relate to PR #1943?
> I mean, I think there will be conflicts whichever order this one and that one are merged.

On the paper it looks like some adjustment would be needed in `PriorityField.ts`. But I think that it is the tests that should have the final word here.

> From the PR description it's not obvious if this is talking about renaming the priorities themselves. Something more specific would be really useful.

Indeed! Fixed.

> If this one is going to go ahead, please do update the user docs, so that a complete PR can be reviewed. Many thanks.

Yup, after I changed the name of the PR, the update of the docs was self evident, thanks =)

claremacrae commented on src/Query/Filter/Field.ts:
```diff
@@ -360,7 +360,7 @@ export abstract class Field {
      */
     public createGrouper(reverse: boolean): Grouper {
-        return new Grouper(this.fieldNameSingular(), this.grouper(), reverse);
+        return new Grouper(this.fieldNameSingular(), this.grouper(), reverse, this.comparator());
```
I thing this will change (improve) the sort order for urgency grouping.

Which will just require confirmation and then a change the docs of group by urgency please.

ilandikov replied:
I checked the docs of urgency [here](https://publish.obsidian.md/tasks/Queries/Grouping#Other+task+properties) and [here](https://publish.obsidian.md/tasks/Advanced/Urgency) but didn't find it related to the changes in this PR (Priority group names...).

Do you mean that now sorting of groups from urgency field can use this code? I agree with that, but I guess this should be a different PR.

claremacrae replied:
It’s this text:

> - Currently, the groups run from the lowest urgency to highest.
> - You can reverse this with group by urgency reverse.
> - In a future release, the default group order will become from the highest urgency to lowest.

I am reasonably confident that this PR has done point 3 and the above text should be updated.

claremacrae replied:
> didn't find it related to the changes in this PR (Priority group names...).

To be more specific, this PR does several things:

1. change the **naming** of groups for the Priority field
2. change the sort of order of grouping for all fields to **sort by task sorting order** instead of by group name alphabetical order

ilandikov replied:
You are right! I wrote a test and tried it before and after all the changes and it does change the behaviour indeed. [Test](https://github.com/obsidian-tasks-group/obsidian-tasks/pull/1966/commits/b7f9f2025237bc83d7e142beafc27d153538e8b8) and [docs](https://github.com/obsidian-tasks-group/obsidian-tasks/pull/1966/commits/67a5baff8c65a545f31891fc07ca2c300ddb53bb).

Now I'm thinking that I should've probably split the PR in 2 - one for refactoring with addition of the comparator to the grouper and the other one to fix the Priority group names. May be a third for the Urgency =)

ilandikov replied:
Lmk if you wish to have a separate issue for the reversed urgency order =)

ilandikov commented on src/Query/Filter/MultiTextField.ts:
```diff
@@ -64,7 +64,7 @@ export abstract class MultiTextField extends TextField {
      */
     public createGrouper(reverse: boolean): Grouper {
-        return new Grouper(this.fieldNamePlural(), this.grouper(), reverse);
+        return new Grouper(this.fieldNamePlural(), this.grouper(), reverse, this.comparator());
```
> What happens with this code if the field does not implement sorting?
> 
> I expect that calling `this.comparator()` will then throw some kind of exception.

Very good point, thanks.
 
> So the testing trick here is to write a new test and have it use a simple implementation for Field that does grouping but not sorting. Perhaps `DescriptionLengthGroupingfield`. A field that is written purely to make testing easy and self-explanatory.

Yup, I did that now, this seems to be a good trade-off.

> Then line 67 can be changed to only use `this.comparator()` if `supportsSorting()` is true, and if not, fall back on the old behaviour of sorting by group name.

On `Field` level I don't have access to the group names so I can't compare them in the fall back case. I'm not really sure how to solve this.

ilandikov replied:
Ok, I think I got this, but not sure whether it is a good solution. Something smells here, can't tell what exactly. Probably the whole direction of the development ㅠㅠ

claremacrae commented:
> Lmk if you wish to have a separate issue for the reversed urgency order =)

I think what this PR has shown is that:

- When `group by` was first implemented, all the groups were sorted by name. The test requirements were therefore:
    - A very few tests to confirm that groups were sorted correctly by name - this could be done in (IIRC) `Group.test.ts`
    - One or two tests of the group names that were generated for each field.
- Now that per-field sorting of groups is done, there is an obvious gap in the tests:
    - We need a test for each field to show the order in which the group names are sorted...

So I currently see these steps:

### PR 1 On the current code, add an expressive test for every field, showing current headings order

It's important that we find a more expressive way to test the group headings order than the current very verbose group testing.

For this PR, this is the sort of thing I mean by expressive (typed from memory without an IDE)

```typescript
expectGroupHeadingsToBe(
    new PriorityField().grouper(),
    shuffle(SampleTasks.withAllPriorities()),
    [
        'Priority 1: High',
        'Priority 2: Medium',
        'Priority 3: None',
        'Priority 4: Low',
    ]
    );    
```

I am thinking that `SampleTasks` would have static methods that each return an array of Task objects with a representative set of all possible values for a given property.

Examples would be:

- `SampleTasks.withAllPriorities()` would be straightforward
- `SampleTasks.withAllDueDateTypes()` could return tasks with each of:
    - A valid date - today
    - A valid date - in the past
    - A valid date - in the future
    - An invalid date
    - No date
    - (Maybe `SampleTasks.withRepresentativeDueDates()` is a better name)

`expectGroupHeadingsToBe()` could be based on ideas in [expectTaskComparesBefore()](https://github.com/obsidian-tasks-group/obsidian-tasks/blob/main/tests/CustomMatchers/CustomMatchersForSorting.ts)

The `shuffle()` call is a nice-to-have, to randomise the order of tasks, so that the tests do not pass by luck

### PR 2 Implement the new group sorting mechanism

Do not change the names of any groups, but do use the new sorting approach that you are now using in the current PR...

The new tests will confirm whether I am correct - but I **think** you will find that only the urgency order will be changed - and it is an improvement.

If I am correct that only urgency changes, then this can actually be labeled as bug-fix, to group most urgent tasks first... and just update the urgency grouping docs.

### PR 3 Rename the priority groups

Now the priority group names can be changed easily, and the docs updated, as a small PR.

ilandikov commented:
Very clear, thanks a lot! I close this PR since we need to split it.

claremacrae commented:
In my example test above I called grouper() but it maybe should have been createNormalGrouper() or something…

ilandikov commented:
> ### PR 1 On the current code, add an expressive test for every field, showing current headings order
> It's important that we find a more expressive way to test the group headings order than the current very verbose group testing.
> 

#1999 
 
> ### PR 2 Implement the new group sorting mechanism
> Do not change the names of any groups, but do use the new sorting approach that you are now using in the current PR...
> 

refactor: #2018

> ### PR 3 Rename the priority groups
> Now the priority group names can be changed easily, and the docs updated, as a small PR.

claremacrae commented:
> > ### PR 1 On the current code, add an expressive test for every field, showing current headings order
> > It's important that we find a more expressive way to test the group headings order than the current very verbose group testing.
> 
> test: #1999

Yeah, sorry I rushed you to create that PR before tests were added for every field.

claremacrae commented:
> Yeah, sorry I rushed you to create that PR before tests were added for every field.

Also, the 'expressive' bit is still important 😄
